README.md
# time_chart


[![pub package](https://img.shields.io/pub/v/time_chart.svg)](https://pub.dev/packages/time_chart)

An amazing time chart in Flutter.

### Chart Types

|                          TimeChart                           |                         AmountChart                          |
| :----------------------------------------------------------: | :----------------------------------------------------------: |
| ![](https://github.com/jja08111/time_chart/blob/main/assets/images/time_chart/weekly_time_chart.gif?raw=true) | ![](https://github.com/jja08111/time_chart/blob/main/assets/images/amount_chart/weekly_amount_chart.gif?raw=true) |
| ![](https://github.com/jja08111/time_chart/blob/main/assets/images/time_chart/monthly_time_chart.gif?raw=true) | ![](https://github.com/jja08111/time_chart/blob/main/assets/images/amount_chart/monthly_amount_chart.gif?raw=true) |



## Getting Started

### 1 - Depend on it

Add it to your package's pubspec.yaml file

```yml
dependencies:
  time_chart: ^0.5.4
```

### 2 - Install it

Install packages from the command line

```sh
flutter packages get
```

### 3 - Usage

Just input your `DateTimeRange` list to `data:` argument. *The list must be sorted.* First data is
latest, last data is oldest.



## Example

```dart
import 'package:flutter/material.dart';
import 'package:time_chart/time_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {

  // Data must be sorted.
  final data = [
    DateTimeRange(
      start: DateTime(2021,2,24,23,15),
      end: DateTime(2021,2,25,7,30),
    ),
    DateTimeRange(
      start: DateTime(2021,2,22,1,55),
      end: DateTime(2021,2,22,9,12),
    ),
    DateTimeRange(
      start: DateTime(2021,2,20,0,25),
      end: DateTime(2021,2,20,7,34),
    ),
    DateTimeRange(
      start: DateTime(2021,2,17,21,23),
      end: DateTime(2021,2,18,4,52),
    ),
    DateTimeRange(
      start: DateTime(2021,2,13,6,32),
      end: DateTime(2021,2,13,13,12),
    ),
    DateTimeRange(
      start: DateTime(2021,2,1,9,32),
      end: DateTime(2021,2,1,15,22),
    ),
    DateTimeRange(
      start: DateTime(2021,1,22,12,10),
      end: DateTime(2021,1,22,16,20),
    ),
  ];

  @override
  Widget build(BuildContext context) {
    final sizedBox = const SizedBox(height: 16);

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('Time chart example app')),
        body: SingleChildScrollView(
          child: Padding(
            padding: const EdgeInsets.all(16.0),
            child: Column(
              children: [
                const Text('Weekly time chart'),
                TimeChart(
                  data: data,
                  viewMode: ViewMode.weekly,
                ),
                sizedBox,
                const Text('Monthly time chart'),
                TimeChart(
                  data: data,
                  viewMode: ViewMode.monthly,
                ),
                sizedBox,
                const Text('Weekly amount chart'),
                TimeChart(
                  data: data,
                  chartType: ChartType.amount,
                  viewMode: ViewMode.weekly,
                  barColor: Colors.deepPurple,
                ),
                sizedBox,
                const Text('Monthly amount chart'),
                TimeChart(
                  data: data,
                  chartType: ChartType.amount,
                  viewMode: ViewMode.monthly,
                  barColor: Colors.deepPurple,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

## Supported languages

|                           English                            |                            Korean                            |
| :----------------------------------------------------------: | :----------------------------------------------------------: |
| ![](https://github.com/jja08111/time_chart/blob/main/assets/images/time_chart/weekly_time_chart.gif?raw=true) | ![](https://github.com/jja08111/time_chart/blob/main/assets/images/time_chart/weekly_time_chart_ko.gif?raw=true) |

You can also use korean language by [Internationalizing Flutter apps](https://flutter.dev/docs/development/accessibility-and-localization/internationalization).


CHANGELOG.md
## 0.5.5
* [Bug] Fixed dart analysis error and warnings.

## 0.5.4
* [Update] Fixed the pub score by changing the intl version.

## 0.5.3
* [Update] Updated library versions.

## 0.5.2
* [Update] Updated y labels of the amount chart. ([#30](https://github.com/jja08111/time_chart/pull/30))

## 0.5.1
* [But] Fixed bug that chart is not updating when data is replaced. ([#28](https://github.com/jja08111/time_chart/pull/28))

## 0.5.0 
* [Breaking] Replaced the `getViewModeLimitDay` function to the `dayCount` enum getter.
* [Update] Refactor the `ChartEngine` class. 

## 0.4.0
* [Update] Migrate Flutter version to 3.0.0 

## 0.3.4
* [Bug] Fixed the incorrect GestureDetector arguments error.

## 0.3.3
* [Update] Clean and improve code.
* [Update] Removed unnecessary data processing.

## 0.3.2
* [Update] Lint code and migrate example app.

## 0.3.1
* [Bug] Fixed a bug if the data list is empty.

## 0.3.0+1
* [Update] Updated path_drawing package version to 1.0.0

## 0.3.0
* [Update] Added a `defaultPivotHour` parameter.
* [Update] Change to use default value of `viewMode`.
* [Bug] Fixed and simplified TimeDataProcessor.

## 0.2.10
* [Bug] Fixed max scroll position bug in scroll physics.

## 0.2.9
* [Bug] Fixed over scrolling bug in IOS.
* [Update] Simplify custom scroll physics.

## 0.2.8
* [Bug] Fixed overflow caused from big text height.
* [Update] Added a `activeTooltip` parameter.
* [Update] Added time chart widget tests.

## 0.2.7
* [Bug] Fixed state bug when changing `ViewMode` in `StatefulWidget`.

## 0.2.6
* [Invalid] Fixed comments
* [Update] Simplify min top position of tooltip

## 0.2.5
* [Bug] Fixed chart state bug caused LayoutBuilder.

## 0.2.4
* [Bug] Fixed invalid sum in amount chart.
* [Bug] Fixed chart layout bug.

## 0.2.3
* [Update] Improved performance by using binary search.

## 0.2.2
* [Update] Code formatting to fix pub score.

## 0.2.1
* [Update] Change how to repaint chart when scrolling by using ValueNotifier.

## 0.2.0
* [Update] Improved performance ([#4](https://github.com/jja08111/time_chart/pull/4)).

## 0.1.0
* [Update] Stable release for null safety.

## 0.1.0-nullsafety.1
* [Update] Modified to use touchable plugin because
[PR](https://github.com/nateshmbhat/touchable/pull/17) was merged.

## 0.1.0-nullsafety.0
* [Update] Null safe migration.

## 0.0.8
* [Update] Code was formatted.


## 0.0.7
* [Update] Added timeChartSizeAnimationDuration property.


## 0.0.6
* [Update] Updated size animation.
* [Bug] Fixed tooltip so that it doesn't change the size by textScaleFactor.


## 0.0.5
* [Update] Migrated Flutter version.


## 0.0.4
* [Typo] Fixed pub score.


## 0.0.3
* [Bug] Fixed invalid range reference.


## 0.0.2
* [Stable] Change sdk version to `>=2.5.2 <3.0.0`.


## 0.0.1 Released on 2021 Feb 25
* Released on.


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:time_chart/time_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  MyApp({Key? key}) : super(key: key);

  // Data must be sorted.
  final List<DateTimeRange> smallDataList = [
    DateTimeRangeWithColor(
      start: DateTime(2021, 2, 24, 23, 15),
      end: DateTime(2021, 2, 25, 7, 30),
      color: Colors.green,
    ),
    DateTimeRangeWithColor(
      color: Colors.red,
      start: DateTime(2021, 2, 22, 1, 55),
      end: DateTime(2021, 2, 22, 9, 12),
    ),
    DateTimeRange(
      start: DateTime(2021, 2, 20, 0, 25),
      end: DateTime(2021, 2, 20, 7, 34),
    ),
    DateTimeRange(
      start: DateTime(2021, 2, 17, 21, 23),
      end: DateTime(2021, 2, 18, 4, 52),
    ),
    DateTimeRange(
      start: DateTime(2021, 2, 13, 6, 32),
      end: DateTime(2021, 2, 13, 13, 12),
    ),
    DateTimeRange(
      start: DateTime(2021, 2, 1, 9, 32),
      end: DateTime(2021, 2, 1, 15, 22),
    ),
    DateTimeRangeWithColor(
      color: Colors.yellow,
      start: DateTime(2021, 1, 22, 12, 10),
      end: DateTime(2021, 1, 22, 16, 20),
    ),
  ];

  List<DateTimeRange> getRandomSampleDataList() {
    final List<DateTimeRange> list = [];
    final random = Random();

    for (int i = 0; i < 10000; ++i) {
      final int randomMinutes1 = random.nextInt(59);
      final int randomMinutes2 = random.nextInt(59);
      final start = DateTime(2021, 2, 1 - i, 0, randomMinutes1);
      final end = DateTime(2021, 2, 1 - i, 7, randomMinutes2 + randomMinutes1);

      list.add(DateTimeRange(
        start: start,
        end: end,
      ));
    }
    return list;
  }

  late final List<DateTimeRange> bigDataList = getRandomSampleDataList();

  @override
  Widget build(BuildContext context) {
    const sizedBox = SizedBox(height: 16);

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('Time chart example app')),
        body: SingleChildScrollView(
          child: Padding(
            padding: const EdgeInsets.all(16.0),
            child: Column(
              children: [
                const Text('Weekly time chart'),
                TimeChart(
                  data: bigDataList,
                  viewMode: ViewMode.weekly,
                ),
                sizedBox,
                const Text('Monthly time chart'),
                TimeChart(
                  data: bigDataList,
                  viewMode: ViewMode.monthly,
                ),
                sizedBox,
                const Text('Weekly amount chart'),
                TimeChart(
                  data: smallDataList,
                  viewMode: ViewMode.weekly,
                  barColor: Colors.deepPurple,
                ),
                sizedBox,
                const Text('Monthly amount chart'),
                TimeChart(
                  data: smallDataList,
                  chartType: ChartType.amount,
                  viewMode: ViewMode.monthly,
                  barColor: Colors.deepPurple,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
